File: sdk/php/tests/e2e/rest/CopyTrading/FuturesTest.php
<?php

namespace Tests\e2e\rest\CopyTrading;

use JMS\Serializer\Serializer;
use JMS\Serializer\SerializerBuilder;
use KuCoin\UniversalSDK\Api\DefaultClient;
use KuCoin\UniversalSDK\Common\Logger;
use KuCoin\UniversalSDK\Generate\CopyTrading\Futures\AddIsolatedMarginReq;
use KuCoin\UniversalSDK\Generate\CopyTrading\Futures\AddOrderReq;
use KuCoin\UniversalSDK\Generate\CopyTrading\Futures\AddOrderTestReq;
use KuCoin\UniversalSDK\Generate\CopyTrading\Futures\AddTPSLOrderReq;
use KuCoin\UniversalSDK\Generate\CopyTrading\Futures\CancelOrderByClientOidReq;
use KuCoin\UniversalSDK\Generate\CopyTrading\Futures\CancelOrderByIdReq;
use KuCoin\UniversalSDK\Generate\CopyTrading\Futures\FuturesApi;
use KuCoin\UniversalSDK\Generate\CopyTrading\Futures\GetMaxOpenSizeReq;
use KuCoin\UniversalSDK\Generate\CopyTrading\Futures\GetMaxWithdrawMarginReq;
use KuCoin\UniversalSDK\Generate\CopyTrading\Futures\ModifyAutoDepositStatusReq;
use KuCoin\UniversalSDK\Generate\CopyTrading\Futures\ModifyIsolatedMarginRiskLimtReq;
use KuCoin\UniversalSDK\Generate\CopyTrading\Futures\RemoveIsolatedMarginReq;
use KuCoin\UniversalSDK\Internal\Utils\JsonSerializedHandler;
use KuCoin\UniversalSDK\Model\ClientOptionBuilder;
use KuCoin\UniversalSDK\Model\Constants;
use KuCoin\UniversalSDK\Model\TransportOptionBuilder;
use PHPUnit\Framework\TestCase;
use Ramsey\Uuid\Uuid;

class FuturesTest extends TestCase
{
    /**
     * @var FuturesApi $api
     */
    private $api;

    /**
     * @var Serializer $serializer
     */
    private $serializer;

    protected function setUp(): void
    {

        $this->serializer = SerializerBuilder::create()
            ->addDefaultHandlers()
            ->configureHandlers(function ($handlerRegistry) {
                $handlerRegistry->registerSubscribingHandler(
                    new JsonSerializedHandler()
                );
            })
            ->build();

        // Retrieve API secret information from environment variables
        $key = getenv('API_KEY') ?: '';
        $secret = getenv('API_SECRET') ?: '';
        $passphrase = getenv('API_PASSPHRASE') ?: '';

        // Optional: Retrieve broker secret information from environment variables; applicable for broker API only
        $brokerName = getenv('BROKER_NAME');
        $brokerKey = getenv('BROKER_KEY');
        $brokerPartner = getenv('BROKER_PARTNER');

        // Set specific options, others will fall back to default values
        $httpTransportOption = (new TransportOptionBuilder())
            ->setKeepAlive(true)
            ->setMaxConnections(10)
            ->build();

        // Create a client using the specified options
        $clientOption = (new ClientOptionBuilder())
            ->setKey($key)
            ->setSecret($secret)
            ->setPassphrase($passphrase)
            ->setBrokerName($brokerName)
            ->setBrokerKey($brokerKey)
            ->setBrokerPartner($brokerPartner)
            ->setSpotEndpoint(Constants::GLOBAL_API_ENDPOINT)
            ->setFuturesEndpoint(Constants::GLOBAL_FUTURES_API_ENDPOINT)
            ->setBrokerEndpoint(Constants::GLOBAL_BROKER_API_ENDPOINT)
            ->setTransportOption($httpTransportOption)
            ->build();

        $client = new DefaultClient($clientOption);
        $kucoinRestService = $client->restService();
        $this->api = $kucoinRestService->getCopytradingService()->getFuturesApi();
    }


    /**
     * addOrder
     * Add Order
     * /api/v1/copy-trade/futures/orders
     */
    public function testAddOrder()
    {
        $builder = AddOrderReq::builder();
        $builder->setClientOid(Uuid::uuid4()->toString())->setSide("buy")->setSymbol("XBTUSDTM")->
        setLeverage(3.0)->setType("limit")->setMarginMode("ISOLATED")
            ->setPrice("1")->setSize(1);
        $req = $builder->build();
        $resp = $this->api->addOrder($req);
        self::assertNotNull($resp->orderId);
        self::assertNotNull($resp->clientOid);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * addOrderTest
     * Add Order Test
     * /api/v1/copy-trade/futures/orders/test
     */
    public function testAddOrderTest()
    {
        $builder = AddOrderTestReq::builder();
        $builder->setClientOid(Uuid::uuid4()->toString())->setSide("buy")->setSymbol("XBTUSDTM")->
        setLeverage(3.0)->setType("limit")->setMarginMode("ISOLATED")
            ->setPrice("1")->setSize(1);
        $req = $builder->build();
        $resp = $this->api->addOrderTest($req);
        self::assertNotNull($resp->orderId);
        self::assertNotNull($resp->clientOid);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * addTPSLOrder
     * Add Take Profit And Stop Loss Order
     * /api/v1/copy-trade/futures/st-orders
     */
    public function testAddTPSLOrder()
    {
        $builder = AddTPSLOrderReq::builder();
        $builder->setClientOid(Uuid::uuid4()->toString())->setSide("buy")->setSymbol("XBTUSDTM")->
        setLeverage(3.0)->setType("limit")->setMarginMode("ISOLATED")->setStopPriceType("TP")
            ->setPrice("10000")->setTriggerStopUpPrice("8000")->setTriggerStopDownPrice("12000")->setSize(1);
        $req = $builder->build();
        $resp = $this->api->addTPSLOrder($req);
        self::assertNotNull($resp->orderId);
        self::assertNotNull($resp->clientOid);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * cancelOrderById
     * Cancel Order By OrderId
     * /api/v1/copy-trade/futures/orders
     */
    public function testCancelOrderById()
    {
        $builder = CancelOrderByIdReq::builder();
        $builder->setOrderId("312410953021030400");
        $req = $builder->build();
        $resp = $this->api->cancelOrderById($req);
        foreach ($resp->cancelledOrderIds as $item) {
            self::assertNotNull($item);
        }

        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * cancelOrderByClientOid
     * Cancel Order By ClientOid
     * /api/v1/copy-trade/futures/orders/client-order
     */
    public function testCancelOrderByClientOid()
    {
        $builder = CancelOrderByClientOidReq::builder();
        $builder->setSymbol("XBTUSDTM")->setClientOid("bfccb19c-55d7-4b28-bcf6-09abf80fa26d");
        $req = $builder->build();
        $resp = $this->api->cancelOrderByClientOid($req);
        self::assertNotNull($resp->clientOid);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * getMaxOpenSize
     * Get Max Open Size
     * /api/v1/copy-trade/futures/get-max-open-size
     */
    public function testGetMaxOpenSize()
    {
        $builder = GetMaxOpenSizeReq::builder();
        $builder->setSymbol("XBTUSDTM")->setPrice(0.1)->setLeverage(10);
        $req = $builder->build();
        $resp = $this->api->getMaxOpenSize($req);
        self::assertNotNull($resp->symbol);
        self::assertNotNull($resp->maxBuyOpenSize);
        self::assertNotNull($resp->maxSellOpenSize);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * getMaxWithdrawMargin
     * Get Max Withdraw Margin
     * /api/v1/copy-trade/futures/position/margin/max-withdraw-margin
     */
    public function testGetMaxWithdrawMargin()
    {
        $builder = GetMaxWithdrawMarginReq::builder();
        $builder->setSymbol("XBTUSDTM");
        $req = $builder->build();
        $resp = $this->api->getMaxWithdrawMargin($req);
        self::assertNotNull($resp->data);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * addIsolatedMargin
     * Add Isolated Margin
     * /api/v1/copy-trade/futures/position/margin/deposit-margin
     */
    public function testAddIsolatedMargin()
    {
        $builder = AddIsolatedMarginReq::builder();
        $builder->setSymbol("XBTUSDTM")->setMargin(3)->setBizNo(uuid::uuid4()->toString());
        $req = $builder->build();
        $resp = $this->api->addIsolatedMargin($req);
        self::assertNotNull($resp->id);
        self::assertNotNull($resp->symbol);
        self::assertNotNull($resp->autoDeposit);
        self::assertNotNull($resp->maintMarginReq);
        self::assertNotNull($resp->riskLimit);
        self::assertNotNull($resp->realLeverage);
        self::assertNotNull($resp->crossMode);
        self::assertNotNull($resp->delevPercentage);
        self::assertNotNull($resp->openingTimestamp);
        self::assertNotNull($resp->currentTimestamp);
        self::assertNotNull($resp->currentQty);
        self::assertNotNull($resp->currentCost);
        self::assertNotNull($resp->currentComm);
        self::assertNotNull($resp->unrealisedCost);
        self::assertNotNull($resp->realisedGrossCost);
        self::assertNotNull($resp->realisedCost);
        self::assertNotNull($resp->isOpen);
        self::assertNotNull($resp->markPrice);
        self::assertNotNull($resp->markValue);
        self::assertNotNull($resp->posCost);
        self::assertNotNull($resp->posCross);
        self::assertNotNull($resp->posInit);
        self::assertNotNull($resp->posComm);
        self::assertNotNull($resp->posLoss);
        self::assertNotNull($resp->posMargin);
        self::assertNotNull($resp->posMaint);
        self::assertNotNull($resp->maintMargin);
        self::assertNotNull($resp->realisedGrossPnl);
        self::assertNotNull($resp->realisedPnl);
        self::assertNotNull($resp->unrealisedPnl);
        self::assertNotNull($resp->unrealisedPnlPcnt);
        self::assertNotNull($resp->unrealisedRoePcnt);
        self::assertNotNull($resp->avgEntryPrice);
        self::assertNotNull($resp->liquidationPrice);
        self::assertNotNull($resp->bankruptPrice);
        self::assertNotNull($resp->settleCurrency);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * removeIsolatedMargin
     * Remove Isolated Margin
     * /api/v1/copy-trade/futures/position/margin/withdraw-margin
     */
    public function testRemoveIsolatedMargin()
    {
        $builder = RemoveIsolatedMarginReq::builder();
        $builder->setSymbol("XBTUSDTM")->setWithdrawAmount(0.000001);
        $req = $builder->build();
        $resp = $this->api->removeIsolatedMargin($req);
        self::assertNotNull($resp->data);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * modifyIsolatedMarginRiskLimt
     * Modify Isolated Margin Risk Limit
     * /api/v1/copy-trade/futures/position/risk-limit-level/change
     */
    public function testModifyIsolatedMarginRiskLimt()
    {
        $builder = ModifyIsolatedMarginRiskLimtReq::builder();
        $builder->setSymbol("XBTUSDTM")->setLevel(1);
        $req = $builder->build();
        $resp = $this->api->modifyIsolatedMarginRiskLimt($req);
        self::assertNotNull($resp->data);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * modifyAutoDepositStatus
     * Modify Isolated Margin Auto-Deposit Status
     * /api/v1/copy-trade/futures/position/margin/auto-deposit-status
     */
    public function testModifyAutoDepositStatus()
    {
        $builder = ModifyAutoDepositStatusReq::builder();
        $builder->setSymbol("XBTUSDTM")->setStatus(true);
        $req = $builder->build();
        $resp = $this->api->modifyAutoDepositStatus($req);
        self::assertNotNull($resp->data);
        Logger::info($resp->jsonSerialize($this->serializer));
    }
}

File: sdk/java/src/test/java/com/kucoin/universal/sdk/test/e2e/rest/copytrading/FuturesTest.java
package com.kucoin.universal.sdk.test.e2e.rest.copytrading;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.kucoin.universal.sdk.api.DefaultKucoinClient;
import com.kucoin.universal.sdk.api.KucoinClient;
import com.kucoin.universal.sdk.generate.copytrading.futures.*;
import com.kucoin.universal.sdk.model.ClientOption;
import com.kucoin.universal.sdk.model.Constants;
import com.kucoin.universal.sdk.model.TransportOption;
import java.io.IOException;
import java.util.Collections;
import java.util.UUID;
import lombok.extern.slf4j.Slf4j;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

@Slf4j
public class FuturesTest {

  private static FuturesApi api;

  public static ObjectMapper mapper = new ObjectMapper();

  @BeforeAll
  public static void setUp() {

    String key = System.getenv("API_KEY");
    String secret = System.getenv("API_SECRET");
    String passphrase = System.getenv("API_PASSPHRASE");

    TransportOption httpTransport =
        TransportOption.builder()
            .interceptors(
                Collections.singleton(
                    new Interceptor() {
                      @NotNull @Override
                      public Response intercept(@NotNull Chain chain) throws IOException {
                        Request request = chain.request();

                        System.out.println("========== Request ==========");
                        System.out.println(request.method() + " " + request.url());

                        Response response = chain.proceed(request);

                        System.out.println("========== Response ==========");
                        System.out.println("Status Code: " + response.code());
                        System.out.println("Message: " + response.message());
                        return response;
                      }
                    }))
            .build();

    ClientOption clientOpt =
        ClientOption.builder()
            .key(key)
            .secret(secret)
            .passphrase(passphrase)
            .spotEndpoint(Constants.GLOBAL_API_ENDPOINT)
            .futuresEndpoint(Constants.GLOBAL_FUTURES_API_ENDPOINT)
            .brokerEndpoint(Constants.GLOBAL_BROKER_API_ENDPOINT)
            .transportOption(httpTransport)
            .build();

    KucoinClient kucoinClient = new DefaultKucoinClient(clientOpt);
    api = kucoinClient.getRestService().getCopytradingService().getFuturesApi();
  }

  /** addOrder Add Order /api/v1/copy-trade/futures/orders */
  @Test
  public void testAddOrder() throws Exception {
    AddOrderReq.AddOrderReqBuilder builder = AddOrderReq.builder();
    builder
        .clientOid(UUID.randomUUID().toString())
        .side(AddOrderReq.SideEnum.BUY)
        .symbol("XBTUSDTM")
        .leverage(3)
        .type(AddOrderReq.TypeEnum.LIMIT)
        .reduceOnly(false)
        .marginMode(AddOrderReq.MarginModeEnum.ISOLATED)
        .price("0.1")
        .size(1)
        .timeInForce(AddOrderReq.TimeInForceEnum.GOODTILLCANCELED);
    AddOrderReq req = builder.build();
    AddOrderResp resp = api.addOrder(req);
    Assertions.assertNotNull(resp.getOrderId());
    Assertions.assertNotNull(resp.getClientOid());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** addOrderTest Add Order Test /api/v1/copy-trade/futures/orders/test */
  @Test
  public void testAddOrderTest() throws Exception {
    AddOrderTestReq.AddOrderTestReqBuilder builder = AddOrderTestReq.builder();
    builder
        .clientOid(UUID.randomUUID().toString())
        .side(AddOrderTestReq.SideEnum.BUY)
        .symbol("XBTUSDTM")
        .leverage(3)
        .type(AddOrderTestReq.TypeEnum.LIMIT)
        .reduceOnly(false)
        .marginMode(AddOrderTestReq.MarginModeEnum.ISOLATED)
        .price("0.1")
        .size(1)
        .timeInForce(AddOrderTestReq.TimeInForceEnum.GOODTILLCANCELED);
    AddOrderTestReq req = builder.build();
    AddOrderTestResp resp = api.addOrderTest(req);
    Assertions.assertNotNull(resp.getOrderId());
    Assertions.assertNotNull(resp.getClientOid());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** addTPSLOrder Add Take Profit And Stop Loss Order /api/v1/copy-trade/futures/st-orders */
  @Test
  public void testAddTPSLOrder() throws Exception {
    AddTPSLOrderReq.AddTPSLOrderReqBuilder builder = AddTPSLOrderReq.builder();
    builder
        .clientOid(UUID.randomUUID().toString())
        .side(AddTPSLOrderReq.SideEnum.BUY)
        .symbol("XBTUSDTM")
        .leverage(3)
        .type(AddTPSLOrderReq.TypeEnum.LIMIT)
        .reduceOnly(false)
        .marginMode(AddTPSLOrderReq.MarginModeEnum.ISOLATED)
        .price("0.1")
        .size(1)
        .timeInForce(AddTPSLOrderReq.TimeInForceEnum.GOODTILLCANCELED)
        .triggerStopUpPrice("0.3")
        .triggerStopDownPrice("0.1")
        .stopPriceType(AddTPSLOrderReq.StopPriceTypeEnum.TRADEPRICE);
    AddTPSLOrderReq req = builder.build();
    AddTPSLOrderResp resp = api.addTPSLOrder(req);
    Assertions.assertNotNull(resp.getOrderId());
    Assertions.assertNotNull(resp.getClientOid());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** cancelOrderById Cancel Order By OrderId /api/v1/copy-trade/futures/orders */
  @Test
  public void testCancelOrderById() throws Exception {
    CancelOrderByIdReq.CancelOrderByIdReqBuilder builder = CancelOrderByIdReq.builder();
    builder.orderId("338035702044397568");
    CancelOrderByIdReq req = builder.build();
    CancelOrderByIdResp resp = api.cancelOrderById(req);
    resp.getCancelledOrderIds().forEach(item -> {});

    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /**
   * cancelOrderByClientOid Cancel Order By ClientOid /api/v1/copy-trade/futures/orders/client-order
   */
  @Test
  public void testCancelOrderByClientOid() throws Exception {
    CancelOrderByClientOidReq.CancelOrderByClientOidReqBuilder builder =
        CancelOrderByClientOidReq.builder();
    builder.symbol("XBTUSDTM").clientOid("465e8c0f-6026-4331-ae8f-2c75aaf1370f");
    CancelOrderByClientOidReq req = builder.build();
    CancelOrderByClientOidResp resp = api.cancelOrderByClientOid(req);
    Assertions.assertNotNull(resp.getClientOid());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** getMaxOpenSize Get Max Open Size /api/v1/copy-trade/futures/get-max-open-size */
  @Test
  public void testGetMaxOpenSize() throws Exception {
    GetMaxOpenSizeReq.GetMaxOpenSizeReqBuilder builder = GetMaxOpenSizeReq.builder();
    builder.symbol("XBTUSDTM").price(0.1).leverage(10);
    GetMaxOpenSizeReq req = builder.build();
    GetMaxOpenSizeResp resp = api.getMaxOpenSize(req);
    Assertions.assertNotNull(resp.getSymbol());
    Assertions.assertNotNull(resp.getMaxBuyOpenSize());
    Assertions.assertNotNull(resp.getMaxSellOpenSize());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /**
   * getMaxWithdrawMargin Get Max Withdraw Margin
   * /api/v1/copy-trade/futures/position/margin/max-withdraw-margin
   */
  @Test
  public void testGetMaxWithdrawMargin() throws Exception {
    GetMaxWithdrawMarginReq.GetMaxWithdrawMarginReqBuilder builder =
        GetMaxWithdrawMarginReq.builder();
    builder.symbol("XBTUSDTM");
    GetMaxWithdrawMarginReq req = builder.build();
    GetMaxWithdrawMarginResp resp = api.getMaxWithdrawMargin(req);
    Assertions.assertNotNull(resp.getData());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /**
   * addIsolatedMargin Add Isolated Margin /api/v1/copy-trade/futures/position/margin/deposit-margin
   */
  @Test
  public void testAddIsolatedMargin() throws Exception {
    AddIsolatedMarginReq.AddIsolatedMarginReqBuilder builder = AddIsolatedMarginReq.builder();
    builder.symbol("XBTUSDTM").margin(3.0).bizNo(UUID.randomUUID().toString());
    AddIsolatedMarginReq req = builder.build();
    AddIsolatedMarginResp resp = api.addIsolatedMargin(req);
    Assertions.assertNotNull(resp.getId());
    Assertions.assertNotNull(resp.getSymbol());
    Assertions.assertNotNull(resp.getAutoDeposit());
    Assertions.assertNotNull(resp.getMaintMarginReq());
    Assertions.assertNotNull(resp.getRiskLimit());
    Assertions.assertNotNull(resp.getRealLeverage());
    Assertions.assertNotNull(resp.getCrossMode());
    Assertions.assertNotNull(resp.getDelevPercentage());
    Assertions.assertNotNull(resp.getOpeningTimestamp());
    Assertions.assertNotNull(resp.getCurrentTimestamp());
    Assertions.assertNotNull(resp.getCurrentQty());
    Assertions.assertNotNull(resp.getCurrentCost());
    Assertions.assertNotNull(resp.getCurrentComm());
    Assertions.assertNotNull(resp.getUnrealisedCost());
    Assertions.assertNotNull(resp.getRealisedGrossCost());
    Assertions.assertNotNull(resp.getRealisedCost());
    Assertions.assertNotNull(resp.getIsOpen());
    Assertions.assertNotNull(resp.getMarkPrice());
    Assertions.assertNotNull(resp.getMarkValue());
    Assertions.assertNotNull(resp.getPosCost());
    Assertions.assertNotNull(resp.getPosCross());
    Assertions.assertNotNull(resp.getPosInit());
    Assertions.assertNotNull(resp.getPosComm());
    Assertions.assertNotNull(resp.getPosLoss());
    Assertions.assertNotNull(resp.getPosMargin());
    Assertions.assertNotNull(resp.getPosMaint());
    Assertions.assertNotNull(resp.getMaintMargin());
    Assertions.assertNotNull(resp.getRealisedGrossPnl());
    Assertions.assertNotNull(resp.getRealisedPnl());
    Assertions.assertNotNull(resp.getUnrealisedPnl());
    Assertions.assertNotNull(resp.getUnrealisedPnlPcnt());
    Assertions.assertNotNull(resp.getUnrealisedRoePcnt());
    Assertions.assertNotNull(resp.getAvgEntryPrice());
    Assertions.assertNotNull(resp.getLiquidationPrice());
    Assertions.assertNotNull(resp.getBankruptPrice());
    Assertions.assertNotNull(resp.getSettleCurrency());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /**
   * removeIsolatedMargin Remove Isolated Margin
   * /api/v1/copy-trade/futures/position/margin/withdraw-margin
   */
  @Test
  public void testRemoveIsolatedMargin() throws Exception {
    RemoveIsolatedMarginReq.RemoveIsolatedMarginReqBuilder builder =
        RemoveIsolatedMarginReq.builder();
    builder.symbol("XBTUSDTM").withdrawAmount(0.000001);
    RemoveIsolatedMarginReq req = builder.build();
    RemoveIsolatedMarginResp resp = api.removeIsolatedMargin(req);
    Assertions.assertNotNull(resp.getData());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /**
   * modifyIsolatedMarginRiskLimt Modify Isolated Margin Risk Limit
   * /api/v1/copy-trade/futures/position/risk-limit-level/change
   */
  @Test
  public void testModifyIsolatedMarginRiskLimt() throws Exception {
    ModifyIsolatedMarginRiskLimtReq.ModifyIsolatedMarginRiskLimtReqBuilder builder =
        ModifyIsolatedMarginRiskLimtReq.builder();
    builder.symbol("XBTUSDTM").level(1);
    ModifyIsolatedMarginRiskLimtReq req = builder.build();
    ModifyIsolatedMarginRiskLimtResp resp = api.modifyIsolatedMarginRiskLimt(req);
    Assertions.assertNotNull(resp.getData());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /**
   * modifyAutoDepositStatus Modify Isolated Margin Auto-Deposit Status
   * /api/v1/copy-trade/futures/position/margin/auto-deposit-status
   */
  @Test
  public void testModifyAutoDepositStatus() throws Exception {
    ModifyAutoDepositStatusReq.ModifyAutoDepositStatusReqBuilder builder =
        ModifyAutoDepositStatusReq.builder();
    builder.symbol("XBTUSDTM").status(true);
    ModifyAutoDepositStatusReq req = builder.build();
    ModifyAutoDepositStatusResp resp = api.modifyAutoDepositStatus(req);
    Assertions.assertNotNull(resp.getData());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }
}
